overlord/configstate/configcore/kernel.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/overlord/devicestate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+const (
+	OptionKernelCmdlineAppend              = "system.kernel.cmdline-append"
+	OptionKernelDangerousCmdlineAppend     = "system.kernel.dangerous-cmdline-append"
+	CoreOptionKernelCmdlineAppend          = "core.system.kernel.cmdline-append"
+	CoreOptionKernelDangerousCmdlineAppend = "core.system.kernel.dangerous-cmdline-append"
+)
+
+func init() {
+	supportedConfigurations[CoreOptionKernelCmdlineAppend] = true
+	supportedConfigurations[CoreOptionKernelDangerousCmdlineAppend] = true
+}
+
+func changedKernelConfigs(c RunTransaction) []string {
+	changed := []string{}
+	for _, name := range c.Changes() {
+		// Note that we cannot just check the prefix as we have
+		// system.kernel.* options also defined in sysctl.go.
+		if name == CoreOptionKernelCmdlineAppend || name == CoreOptionKernelDangerousCmdlineAppend {
+			nameWithoutSnap := strings.SplitN(name, ".", 2)[1]
+			changed = append(changed, nameWithoutSnap)
+		}
+	}
+	return changed
+}
+
+func validateParamsAreAllowed(st *state.State, devCtx snapstate.DeviceContext, params []string) error {
+	// gd, err := devicestate.CurrentGadgetInfo(st, devCtx)
+	// if err != nil {
+	// 	return err
+	// }
+	// logger.Debugf("gadget data read from %s", gd.RootDir)
+	// TODO use gadgetdata to check against allowed values
+
+	return nil
+}
+
+func validateCmdlineExtra(c RunTransaction) error {
+	changed := changedKernelConfigs(c)
+	if len(changed) == 0 {
+		return nil
+	}
+
+	st := c.State()
+	st.Lock()
+	defer st.Unlock()
+	devCtx, err := devicestate.DeviceCtx(st, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	for _, opt := range changed {
+		cmdExtra, err := coreCfg(c, opt)
+		if err != nil {
+			return err
+		}
+
+		logger.Debugf("validating %s=%q", opt, cmdExtra)
+		params, err := osutil.KernelCommandLineSplit(cmdExtra)
+		if err != nil {
+			return err
+		}
+		if opt == OptionKernelCmdlineAppend {
+			// check against allowed values from gadget
+			if err := validateParamsAreAllowed(c.State(), devCtx, params); err != nil {
+				return fmt.Errorf("while validating params: %v", err)
+			}
+		} else { // OptionKernelDangerousCmdlineAppend
+			if devCtx.Model().Grade() != asserts.ModelDangerous {
+				return fmt.Errorf("cannot use %s for non-dangerous model", opt)
+			}
+		}
+	}
+
+	return nil
+}
+
+func handleCmdlineExtra(c RunTransaction, opts *fsOnlyContext) error {
+	kernelOpts := changedKernelConfigs(c)
+	if len(kernelOpts) == 0 {
+		return nil
+	}
+	logger.Debugf("handling %v", kernelOpts)
+
+	st := c.State()
+	st.Lock()
+
+	// error out if some other change is touching the kernel command line
+	if err := snapstate.CheckSetKernelCmdlineConflict(st, ""); err != nil {
+		st.Unlock()
```

Comment:
it would be better to find a refomurlation maybe using helpers that doesn't need this style, also we need a test that reaches here

Reply:
Changed, test added too